File: gdoc-design/gdoc.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

// abstract class

class DocumentElement
{
public:
    virtual string render() const = 0; // Pure virtual function for rendering
};

class textElement : public DocumentElement
{
private:
    string text;

public:
    textElement(const string &text) : text(text) {};

    string render() const override
    {
        return text;
    };
};

class imageElement : public DocumentElement
{
private:
    string imagePath;

public:
    imageElement(const string &path) : imagePath(path) {};

    string render() const override
    {
        return "<img src=\"" + imagePath + "\">";
    }
};

class Document
{
private:
    vector<DocumentElement *> elements;

public:
    void addElement(DocumentElement *element)
    {
        elements.push_back(element);
    };

    string render()
    {
        string result;

        for (const auto &element : elements)
        {
            result += element->render() + "\n";
        }

        return result;
    }
};

class DocumentEditor
{
private:
    Document *document;
    string renderedDocument;

public:
    DocumentEditor(Document *doc) : document(doc) {};

    void addText(const string &text)
    {
        document->addElement(new textElement(text));
    }

    void addImage(const string &imagePath)
    {
        document->addElement(new imageElement(imagePath));
    }

    string renderDocument()
    {
        if (renderedDocument.empty())
        {
            renderedDocument = document->render();
        }

        return renderedDocument;
    }
};

int main(void)
{
    Document *doc = new Document();
    DocumentEditor *editor = new DocumentEditor(doc);
    editor->addText("This is a sample document.");
    editor->addImage("image1.jpg");
    editor->addText("This document contains an image.");
    cout << editor->renderDocument() << endl;
    return 0;
}

File: gdoc-design/gdoc.ts
abstract class documentElement {
    abstract render(): string;
};

class textElement extends documentElement {
    private text: string;
    constructor(text: string) {
        super();
        this.text = text;
    }

    render(): string {
        return this.text;
    }
};


class imageElement extends documentElement {
    private path: string;

    constructor(path: string) {
        super();
        this.path = path;
    };

    render(): string {
        return this.path;
    }
};

class gdocument {
    private element: documentElement[]
        = [];

    addElement(element: documentElement): void {
        this.element.push(element);
    };

    render(): string {
        let result = "";

        for (const ele of this.element) {
            result += ele.render() + '\n';
        }

        return result;
    };
}

class googleDoc {
    private gdoc: gdocument;

    constructor(doc: gdocument) {
        this.gdoc = doc;
    };


    addText(text: string): void {
        this.gdoc.addElement(new textElement(text));
    };


    addImage(url: string): void {
        this.gdoc.addElement(new imageElement(url));
    };

    render(): string {
        return this.gdoc.render();
    };

};

const myDoc = new gdocument();
const googleDocument = new googleDoc(myDoc);

googleDocument.addText("Hello, this is a sample text. ");
googleDocument.addImage("https://example.com/image.png");
googleDocument.addText("This is another text after the image.");
console.log(googleDocument.render())